Grails 5.0.0.M1 - Fail to run application with error NoUniqueBeanDefinitionException for PlatformTransactionManager

_Originally reported by [Jesse Elliott](https://github.com/jjelliott) on the Grails slack_

In Grails 5, we changed the bean autowire from autowire-by-name to autowire-by-type as default which results in the `org.springframework.beans.factory.NoUniqueBeanDefinitionException` when multiple data-sources are configured because there are multiple `PlatformTransactionManager` bean.


**Steps to Reproduce**
1. Create a new web application using `grails create-app example.demo`
2. Create domain class `Institution` as:
```
package example

class Institution {
  String name
}
```
3. Configure multiple data-sources in the application configuration file:
_/grails-app/config/application.yml_
```
dataSource:
    pooled: true
    jmxExport: true
    driverClassName: org.h2.Driver
    dbCreate: create-drop
    url: jdbc:h2:mem:devDb;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE
dataSources:
    secondary:
        pooled: true
        jmxExport: true
        driverClassName: org.h2.Driver
        dbCreate: create-drop
        url: jdbc:h2:mem:devDb2;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE
```

4. Create a data-service for the `Institution` domain as:
```
package example

import grails.gorm.services.Service
import grails.gorm.transactions.ReadOnly

@Service(Institution)
abstract class InstitutionService {

    @ReadOnly
    abstract Institution findByName()

    @ReadOnly
    List<Institution> findAll() {
        Institution.getAll()
    }
}
```
5. Run the application using command `./gradlew bootRun`.

**Expected Behavior**

The application should run fine with `InstitutionService` correctly configured to use the default `PlatformTranasactionManager` bean.

**Actual Behavior**

The application fails to start with the following error:

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'institutionService': Invocation of init method failed; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'example.$InstitutionServiceImplementation': Unsatisfied dependency expressed through bean property 'transactionManager'; nested exception is org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type 'org.springframework.transaction.PlatformTransactionManager' available: expected single matching bean but found 2: transactionManager,transactionManager_secondary
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1788)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:609)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:531)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:925)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:923)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:588)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:767)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:426)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:326)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:99)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:485)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:472)
	at example.Application.main(Application.groovy:11)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'example.$InstitutionServiceImplementation': Unsatisfied dependency expressed through bean property 'transactionManager'; nested exception is org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type 'org.springframework.transaction.PlatformTransactionManager' available: expected single matching bean but found 2: transactionManager,transactionManager_secondary
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireByType(AbstractAutowireCapableBeanFactory.java:1518)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1401)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireBeanProperties(AbstractAutowireCapableBeanFactory.java:406)
	at org.grails.datastore.mapping.config.DatastoreServiceMethodInvokingFactoryBean.invokeWithTargetException(DatastoreServiceMethodInvokingFactoryBean.groovy:37)
	at org.springframework.beans.factory.config.MethodInvokingFactoryBean.afterPropertiesSet(MethodInvokingFactoryBean.java:108)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1847)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1784)
	... 23 common frames omitted
Caused by: org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type 'org.springframework.transaction.PlatformTransactionManager' available: expected single matching bean but found 2: transactionManager,transactionManager_secondary
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveNotUnique(DependencyDescriptor.java:220)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1345)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1287)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireByType(AbstractAutowireCapableBeanFactory.java:1503)
	... 29 common frames omitted
```` 

**Context Information**
* macOS Big Sur (Version 11.1)
* Grails 5.0.0.M1
* Java: openjdk version "1.8.0_252"

Example Application:
https://github.com/jjelliott/grails-5-abstract-data-services

--- a/examples/grails3-multiple-datasources/grails-app/services/example/BookService.groovy
+++ b/examples/grails3-multiple-datasources/grails-app/services/example/BookService.groovy
@@ -1,0 +1,16 @@
+package example
+
+import grails.gorm.services.Service
+import grails.gorm.transactions.ReadOnly
+
+@Service(Book)
+abstract class BookService {
+
+    @ReadOnly
+    abstract Book findByTitle(String title)
+
+    @ReadOnly
+    List<Book> findAll() {
+        Book.getAll()
+    }
+}

--- a/examples/grails3-multiple-datasources/src/integration-test/groovy/functionaltests/MultipleDataSourcesSpec.groovy
+++ b/examples/grails3-multiple-datasources/src/integration-test/groovy/functionaltests/MultipleDataSourcesSpec.groovy
@@ -1,6 +1,7 @@
 package functionaltests
 
 import datasources.Application
+import example.BookService
 import grails.testing.mixin.integration.Integration
 import grails.gorm.transactions.*
 import spock.lang.*
@@ -10,6 +11,8 @@ import ds2.Book as SecondBook
 @Integration(applicationClass = Application)
 @Rollback
 class MultipleDataSourcesSpec extends Specification {
+
+    BookService bookService
 
     void "Test multiple data source persistence"() {
         when:
@@ -23,5 +26,10 @@ class MultipleDataSourcesSpec extends Specification {
             Book.count() == 2
             SecondBook.withTransaction(readOnly: true) { SecondBook.count() } == 1
             SecondBook.withTransaction(readOnly: true)  { SecondBook.secondary.count() } == 1
+    }
+
+    void "test BookService does NOT throw NoUniqueBeanDefinitionException when multiple dataSources are configured"() {
+        expect:
+        bookService != null
     }
 }

--- a/grails-plugin/src/main/groovy/grails/orm/bootstrap/HibernateDatastoreSpringInitializer.groovy
+++ b/grails-plugin/src/main/groovy/grails/orm/bootstrap/HibernateDatastoreSpringInitializer.groovy
@@ -154,8 +154,12 @@ class HibernateDatastoreSpringInitializer extends AbstractDatastoreInitializer {
             hibernateDatastore(HibernateDatastore, config, hibernateConnectionSourceFactory, eventPublisher) { bean->
                 bean.primary = true
             }
-            sessionFactory(hibernateDatastore:'getSessionFactory')
-            transactionManager(hibernateDatastore:"getTransactionManager")
+            sessionFactory(hibernateDatastore:'getSessionFactory') { bean->
+                bean.primary = true
+            }
+            transactionManager(hibernateDatastore:"getTransactionManager") { bean->
+                bean.primary = true
+            }
             autoTimestampEventListener(hibernateDatastore:"getAutoTimestampEventListener")
             getBeanDefinition("transactionManager").beanClass = PlatformTransactionManager
             hibernateDatastoreConnectionSourcesRegistrar(HibernateDatastoreConnectionSourcesRegistrar, dataSources)

